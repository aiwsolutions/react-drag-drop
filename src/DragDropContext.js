/**
 * Copyright by AIWSolutions.
 */
import { findDOMNode } from 'react-dom';
import rbush from 'rbush';
import _ from 'lodash';
import { isTouchable, lockDownSelection, getScrollElement, getDirection, getRectData } from './utils';

class DragDropContext {
    boundaries = rbush();
    idBoundaryMap = {};
    cache = {};
    scrollSensitive = 100;
    scrollSpeed = 10;
    enabled = true;

    getCurrentSource() {
        return this.currentSource;
    }

    getCurrentTarget() {
        return this.currentTarget;
    }

    getCacheBoundary() {
        return this.cache.boundary;
    }

    setEnabled(enabled) {
        this.enabled = enabled;
    }

    isEnabled() {
        return this.enabled;
    }

    removeBoundaryOf(uniqueId) {
        const existing = this.idBoundaryMap[uniqueId];
        if (!existing) {
            return;
        }
        this.boundaries.remove(existing, (a, b) => a.uniqueId === b.uniqueId);
        _.unset(this.idBoundaryMap, uniqueId);
    }

    addBoundary(uniqueId, boundary, extraData) {
        const rectData = getRectData(uniqueId, boundary, extraData);
        this.boundaries.insert(rectData);
        this.idBoundaryMap[uniqueId] = rectData;
    }

    getBoundary(uniqueId) {
        return this.idBoundaryMap[uniqueId];
    }

    getOverredBoundaries() {
        return this.cache.overredBoundaries;
    }

    updateDraggable(source, boundary) {
        const { props: { uniqueId, type } } = source;
        if (!uniqueId) {
            throw new Error('A Draggable must be provided with a unique id');
        }
        this.removeBoundaryOf(uniqueId);
        this.addBoundary(uniqueId, boundary, {
            draggable: true,
            type
        });
        if (this.currentSource && this.currentSource.props.uniqueId === uniqueId) {
            this.currentSource = source;
        }
    }

    unmountDraggable({ props: { uniqueId } }) {
        this.removeBoundaryOf(uniqueId);
    }

    updateDroppable(target, boundary, onDraggedSourceOverred, onDraggedSourceOuted) {
        const { props: { uniqueId, acceptTypes } } = target;
        if (!uniqueId) {
            throw new Error('A Droppable must be provided with a unique id');
        }
        this.removeBoundaryOf(uniqueId);
        this.addBoundary(uniqueId, boundary, {
            droppable: true,
            acceptTypes,
            onDraggedSourceOverred,
            onDraggedSourceOuted
        });
        if (this.currentTarget && this.currentTarget.uniqueId === uniqueId) {
            this.currentTarget = this.getBoundary(uniqueId);
        }
    }

    unmountDroppable({ props: { uniqueId } }) {
        this.removeBoundaryOf(uniqueId);
    }

    shouldContextHandleEvent() {
        return this.cache.dragging;
    }

    startDragging(source, { lastX, lastY, data }) {
        if (!this.enabled) {
            return;
        }
        const { props: { uniqueId } } = source;
        const boundary = this.getBoundary(uniqueId);
        if (!boundary) {
            throw new Error('Unregistered Draggable with uniqueId', uniqueId);
        }
        const { minX, minY } = boundary;
        this.currentSource = source;
        this.cache = {
            offsetX: lastX - minX,
            offsetY: lastY - minY,
            lastX,
            lastY,
            data,
            boundary
        };

        lockDownSelection(true);
        if (isTouchable()) {
            document.addEventListener('touchmove', this.handleTouchDrag, { passive: false, capture: false });
            document.addEventListener('touchend', this.handleTouchDragComplete);
        } else {
            document.addEventListener('mousemove', this.handleDrag);
            document.addEventListener('mouseup', this.handleDragComplete);
        }
    }

    removeEventListeners() {
        if (isTouchable()) {
            document.removeEventListener('touchmove', this.handleTouchDrag);
            document.removeEventListener('touchend', this.handleTouchResize);
        } else {
            document.removeEventListener('mousemove', this.handleDrag);
            document.removeEventListener('mouseup', this.handleDragComplete);
        }
    }

    drawMovingDom() {
        const sourceDOM = findDOMNode(this.currentSource);  // eslint-disable-line
        if (!sourceDOM) {
            throw new Error('Illegal state: dragged source does not exist');
        }

        // notify source that it is being dragged.
        this.currentSource.setDragging(true);

        const {
            boundary: {
                minX, minY, maxX, maxY
            }
        } = this.cache;
        this.movingDOM = sourceDOM.cloneNode(true);
        // if this is fixed, then we have to accumulate scroll position
        this.movingDOM.style.position = 'absolute';
        this.movingDOM.style.width = `${maxX - minX}px`;
        this.movingDOM.style.height = `${maxY - minY}px`;
        this.movingDOM.style.left = `${minX}px`;
        this.movingDOM.style.top = `${minY}px`;
        this.movingDOM.style.zIndex = 10000;
        document.body.appendChild(this.movingDOM);
    }

    moveMovingDomTo({ left, top }) {
        if (!this.movingDOM) {
            return;
        }
        this.movingDOM.style.left = `${left}px`;
        this.movingDOM.style.top = `${top}px`;
        // there is a moment that source is mouting to new place
        if (!this.currentSource) {
            return;
        }
        const { props: { onRendered } } = this.currentSource;
        if (onRendered) {
            const {
                boundary: {
                    minX, maxX, minY, maxY
                }
            } = this.cache;
            const width = maxX - minX;
            const height = maxY - minY;
            onRendered({
                left,
                top,
                right: left + width,
                bottom: top + height,
                width,
                height
            });
        }
    }

    clearMovingDom() {
        if (this.movingDOM) {
            document.body.removeChild(this.movingDOM);
            this.movingDOM = null;
        }
    }

    buildDragEvent({
        clientX, clientY, pageX, pageY
    }) {
        const { lastX, lastY, data } = this.cache;
        const deltaX = pageX - lastX;
        const deltaY = pageY - lastY;

        return {
            clientX,
            clientY,
            pageX,
            pageY,
            deltaX,
            deltaY,
            data,
            directionX: getDirection(lastX, pageX),
            directionY: getDirection(lastY, pageY)
        };
    }

    handleDrag = (e) => {
        if (!this.cache.dragging) {
            this.cache.dragging = true;
        }
        if (!this.movingDOM) {
            this.drawMovingDom();
        }

        this.scrollViewPort(e);
        // set overred boundaries
        this.cache.overredBoundaries = this.boundaries.search({
            minX: e.pageX,
            minY: e.pageY,
            maxX: e.pageX + 1,
            maxY: e.pageY + 1
        });

        const { props: { onDrag, type } } = this.currentSource;
        const dragEvent = this.buildDragEvent(e);
        if (onDrag && !onDrag(dragEvent)) {
            return false;
        }

        const { offsetX, offsetY } = this.cache;
        requestAnimationFrame(() => this.moveMovingDomTo({
            left: e.pageX - offsetX,
            top: e.pageY - offsetY
        }));

        this.findAndNotifyDroppables(type, true);
        _.merge(this.cache, {
            lastX: e.pageX,
            lastY: e.pageY
        });
        return true;
    }

    handleDragComplete = (e) => {
        if (this.currentSource) {
            const { props: { onDrop, type } } = this.currentSource;
            if (onDrop) {
                const dragEvent = this.buildDragEvent(e);
                onDrop(dragEvent);
            }
            this.findAndNotifyDroppables(type, false);
            // notify source that is has been released
            this.currentSource.setDragging(false);
        }
        this.removeEventListeners();
        this.clearMovingDom();
        this.currentSource = null;
        this.currentTarget = null;
        this.cache = {};
        lockDownSelection(false);
        return true;
    }

    handleTouchDrag = (e) => {
        e.preventDefault();
        e.stopPropagation();
        const touch = e.targetTouches[0];
        this.handleDrag(touch);
    }

    handleTouchDragComplete = (e) => {
        const touch = e.changedTouches[0];
        this.handleDragComplete(touch);
    }

    findAndNotifyDroppables(type, dragging) {
        if (!dragging && this.currentTarget) {
            this.currentTarget.onDraggedSourceOuted();
            this.currentTarget = null;
            return;
        }

        _.forEach(this.cache.overredBoundaries, (droppableBoundary) => {
            const {
                uniqueId, droppable, acceptTypes, onDraggedSourceOverred
            } = droppableBoundary;
            if (!droppable || (!_.isEmpty(acceptTypes) && !_.includes(acceptTypes, type))) {
                return true;
            }
            // current target already gets fired with onDraggedSourceOverred
            if (this.currentTarget && this.currentTarget.uniqueId === uniqueId) {
                return false;
            }
            const { data, boundary } = this.cache;
            if (onDraggedSourceOverred({ ...data, ...boundary })) {
                if (this.currentTarget && this.currentTarget.uniqueId !== uniqueId) {
                    this.currentTarget.onDraggedSourceOuted();
                }
                this.currentTarget = droppableBoundary;
                return false;
            }
            return true;
        });
    }

    scrollViewPort({ clientX, clientY }) {
        const viewPortWidth = Math.max(document.documentElement.clientWidth, window.innerWidth || 0);
        const viewPortHeight = Math.max(document.documentElement.clientHeight, window.innerHeight || 0);
        const scrollElement = getScrollElement();
        const scrolledTop = scrollElement.scrollTop;
        const scrolledLeft = scrollElement.scrollLeft;
        if (clientX < this.scrollSensitive && scrolledLeft > 0) {
            scrollElement.scrollLeft = Math.max(0, scrolledLeft - this.scrollSpeed);
        }
        if (clientX > viewPortWidth - this.scrollSensitive) {
            scrollElement.scrollLeft = scrolledLeft + this.scrollSpeed;
        }
        if (clientY < this.scrollSensitive && scrolledTop > 0) {
            scrollElement.scrollTop = Math.max(0, scrolledTop - this.scrollSpeed);
        }
        if (clientY > viewPortHeight - this.scrollSensitive) {
            scrollElement.scrollTop = scrolledTop + this.scrollSpeed;
        }
    }
}

/* eslint no-underscore-dangle: 0 */
if (!window.__iw_drag_drop_context) {
    window.__iw_drag_drop_context = new DragDropContext();
}

export default window.__iw_drag_drop_context;
