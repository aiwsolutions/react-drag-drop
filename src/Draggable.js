/**
 * Copyright by AIWSolutions.
 */
import React from 'react';
import PropTypes from 'prop-types';
import DragDropContext from './DragDropContext';
import { matchesSelector, getComponentBoundary } from './utils';

class Draggable extends React.Component {
    static propTypes = {
        /**
          * Unique id for the component
          */
        uniqueId: PropTypes.any.isRequired,

        /**
         * Start dragging the object.
         * Return the data associated with this element
         * Return false/null to prevent the dragging
         */
        onDragStart: PropTypes.func,

        /**
         * Dragging the object.
         * Return true to accept the move, false to deny
         * Arguments: {deltaX, deltaY}
         */
        onDrag: PropTypes.func,

        /**
         * Stop dragging.
         * Callback only
         */
        onDrop: PropTypes.func,

        /**
         * Call when there is click or touch without dragging or resizing
         */
        onPressed: PropTypes.func,

        /**
         * Callback function for both componentDidUpdate & componentDidMount & dragging
         * with boundary client rect
         */
        onRendered: PropTypes.func,

        /**
         * Callback function for componentWillUnmount
         */
        onUnmount: PropTypes.func,

        /**
         * Skip decorating for selector
         */
        skip: PropTypes.node,

        /**
         * Only decorating for handle
         */
        handle: PropTypes.node,

        /**
         * The type for matching with Droppable
         */
        type: PropTypes.string.isRequired,

        shadowing: PropTypes.bool
    };

    constructor(props) {
        super(props);
        this.state = {
            dragging: DragDropContext.getCurrentSource()
                && DragDropContext.getCurrentSource().props.uniqueId === props.uniqueId
                && DragDropContext.shouldContextHandleEvent()
        };
    }

    componentDidMount() {
        this.updateDraggable();
    }

    componentDidUpdate() {
        this.updateDraggable();
    }

    componentWillUnmount() {
        const { onUnmount } = this.props;
        if (onUnmount) {
            onUnmount();
        }
        DragDropContext.unmountDraggable(this);
    }

    setDragging = dragging => this.setState({ dragging });

    updateDraggable = () => {
        const { onRendered } = this.props;
        const boundary = getComponentBoundary(this);
        DragDropContext.updateDraggable(this, boundary);
        if (onRendered) {
            onRendered(boundary);
        }
    }

    isAcceptedEvent = ({ button, target }) => {
        // left click only
        if (button !== 0) return false;
        if (this.props.disabled) return false;

        // check again for skip && handle
        const { skip, handle } = this.props;
        if (skip && matchesSelector(target, skip)) return false;
        if (handle && !matchesSelector(target, handle)) return false;
        return true;
    }

    handleMouseDown = (event) => {
        if (!DragDropContext.isEnabled() || this.props.disabled || !this.isAcceptedEvent(event)) {
            return false;
        }
        const {
            clientX, clientY, pageX, pageY
        } = event;

        const {
            onDragStart
        } = this.props;

        // if this instance can be dragged or resized, then should not others.
        event.stopPropagation();

        const data = onDragStart({
            clientX, clientY, pageX, pageY
        });

        if (!data) {
            return false;
        }

        DragDropContext.startDragging(this, {
            lastX: pageX,
            lastY: pageY,
            data
        });
        return true;
    };

    handleTouchStart = (e) => {
        const touch = e.nativeEvent.targetTouches[0];
        this.handleMouseDown({
            button: 0,
            clientX: touch.clientX,
            clientY: touch.clientY,
            pageX: touch.pageX,
            pageY: touch.pageY,
            target: touch.target,
            stopPropagation: e.stopPropagation
        });
    };

    handleMouseUp = (e) => {
        if (!DragDropContext.isEnabled() || !this.isAcceptedEvent(e)) {
            return;
        }
        const { onPressed } = this.props;
        if (!DragDropContext.shouldContextHandleEvent() && onPressed) {
            e.stopPropagation();
            onPressed(e);
        }
    };

    handleTouchEnd = (e) => {
        const touch = e.nativeEvent.targetTouches[0];
        this.handleMouseUp({
            ...touch,
            stopPropagation: e.stopPropagation
        });
    }

    handleMovedTo = position => this.dragListener && this.dragListener(position);

    render() {
        const { children } = this.props;
        const { dragging } = this.state;
        if (dragging) {
            const cacheBoundary = DragDropContext.getCacheBoundary();
            // We only draw a placeholder when it is dragging
            if (cacheBoundary) {
                const { minX, maxX, minY, maxY } = cacheBoundary;
                return (
                    <div
                        style={{
                            width: maxX - minX,
                            height: maxY - minY
                        }}
                    />
                );
            }
        }

        const handlers = {
            onTouchStart: this.handleTouchStart,
            onMouseDown: this.handleMouseDown,
            onMouseMove: this.handleMouseMove,
            onMouseUp: this.handleMouseUp,
            onTouchEnd: this.handleTouchEnd
        };

        const renderedChildren = children(handlers);

        return renderedChildren && React.Children.only(renderedChildren);
    }
}

export default Draggable;
