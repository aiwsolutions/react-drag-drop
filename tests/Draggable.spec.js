import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import Draggable from '../src/Draggable';
import DragDropContext from '../src/DragDropContext';

jest.mock('react-dom', () => ({
    findDOMNode: () => ({
        getBoundingClientRect: () => ({
            left: 0, top: 0, right: 0, bottom: 0
        })
    })
}));

function renderSampleComponent() {
    return renderer.create(
        <Draggable
            uniqueId="1"
            type="test"
            onDragStart={f => f}
        >
            {
                handlers =>
                    <div
                        {...handlers}
                        className="test"
                        style={{ backgroundColor: '#0000ff' }}
                    >
                        Sample content
                    </div>
            }
        </Draggable>
    );
}

test('<Draggable> - must render children as it is', () => {
    expect(renderSampleComponent().toJSON()).toMatchSnapshot();
});

test('<Draggable> - when the component is mounted, it updates its boundary to cache', () => {
    sinon.spy(DragDropContext, 'updateDraggable');
    renderSampleComponent();
    expect(DragDropContext.updateDraggable.calledOnce).toBeTruthy();
    DragDropContext.updateDraggable.restore();
});

test('<Draggable> - when the component is updated, it updates its boundary to cache', () => {
    sinon.spy(DragDropContext, 'updateDraggable');

    const comp = renderSampleComponent();
    comp.getInstance().componentDidUpdate();

    expect(DragDropContext.updateDraggable.calledTwice).toBeTruthy();
    DragDropContext.updateDraggable.restore();
});

function buildLeftClickEvent() {
    return {
        button: 0,
        stopPropagation: sinon.stub(),
        pageX: 0,
        pageY: 0
    };
}

test('<Draggable> - when left click on the component, it must call DragDropContext.startDragging', () => {
    sinon.spy(DragDropContext, 'startDragging');

    const comp = renderSampleComponent();
    comp.root.findByType('div').props.onMouseDown(buildLeftClickEvent());

    expect(DragDropContext.startDragging.calledOnce).toBeTruthy();
    expect(DragDropContext.startDragging.getCall(0).args[1])
        .toMatchObject({ lastX: 0, lastY: 0, data: { pageX: 0, pageY: 0 } });
    DragDropContext.startDragging.restore();
});

test('<Draggable> - if onDragStart does not return data, then startDragging should not be called', () => {
    sinon.spy(DragDropContext, 'startDragging');

    const comp = renderer.create(
        <Draggable
            uniqueId={2}
            type="test"
            onDragStart={() => false}
        >
            {
                handlers => <div {...handlers} />
            }
        </Draggable>
    );
    comp.root.findByType('div').props.onMouseDown(buildLeftClickEvent());
    expect(DragDropContext.startDragging.notCalled).toBeTruthy();
});

function renderDraggableWithPressHandler(pressHandler) {
    return renderer.create(
        <Draggable
            uniqueId={3}
            type="test"
            onPressed={pressHandler}
        >
            {handlers => <div {...handlers} />}
        </Draggable>
    );
}

test('<Draggable> - mouse up will call onPressed if the context does not handle drag', () => {
    sinon.stub(DragDropContext, 'shouldContextHandleEvent').returns(false);
    const pressHandler = sinon.stub();

    const comp = renderDraggableWithPressHandler(pressHandler);

    comp.root.findByType('div').props.onMouseUp(buildLeftClickEvent());
    expect(pressHandler.calledOnce).toBeTruthy();
    DragDropContext.shouldContextHandleEvent.restore();
});

test('<Draggable> - mouse up will not call onPressed if there is movement', () => {
    sinon.stub(DragDropContext, 'shouldContextHandleEvent').returns(true);
    const pressHandler = sinon.stub();

    const comp = renderDraggableWithPressHandler(pressHandler);

    comp.root.findByType('div').props.onMouseUp(buildLeftClickEvent());
    expect(pressHandler.notCalled).toBeTruthy();
    DragDropContext.shouldContextHandleEvent.restore();
});
