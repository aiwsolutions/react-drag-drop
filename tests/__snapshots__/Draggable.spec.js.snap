// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Draggable> - must render children as it is 1`] = `
<div
  className="test"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "#0000ff",
    }
  }
>
  Sample content
</div>
`;
